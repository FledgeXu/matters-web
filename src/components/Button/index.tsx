import classNames from 'classnames'
import Link from 'next/link'

import styles from './styles.css'

type ButtonWidth = 'full' | '2rem' | '4rem' | '6rem' | undefined | null

type ButtonHeight =
  | '1rem'
  | '1.25rem'
  | '1.5rem'
  | '2rem'
  | '2.25rem'
  | '3rem'
  | undefined
  | null

type ButtonSpacing =
  | 0
  | '0'
  | 'xxxtight'
  | 'xxtight'
  | 'xtight'
  | 'tight'
  | 'base'
  | 'loose'

type ButtonTextColor = 'white' | 'black' | 'green' | 'red'

type ButtonColor =
  | 'grey'
  | 'grey-lighter'
  | 'green-lighter'
  | 'green'
  | 'gold'
  | 'red'

interface ButtonProps {
  size?: [ButtonWidth, ButtonHeight]
  spacing?: [ButtonSpacing, ButtonSpacing]

  textColor?: ButtonTextColor
  textHoverColor?: ButtonTextColor

  bgColor?: ButtonColor
  bgHoverColor?: ButtonColor

  borderColor?: ButtonColor
  borderWidth?: 'sm'

  style?: React.CSSProperties

  href?: string
  as?: string

  htmlType?: 'button' | 'submit'
  [key: string]: any
}

export const Button: React.FC<ButtonProps> = ({
  spacing = [0, 0],
  size = [null, null],

  textColor,
  textHoverColor,

  bgColor,
  bgHoverColor,

  borderColor,
  borderWidth,

  href,
  as,

  style,
  className,
  htmlType = 'button',
  children,
  ...restProps
}) => {
  const containerClass = classNames({
    container: true,
    [`text-hover-${textHoverColor}`]: !!textHoverColor,
    [`bg-hover-${bgHoverColor}`]: !!bgHoverColor,
    [className]: !!className
  })

  const contentClass = classNames({
    content: true,
    [`width-${size[0]}`]: size[0] === 'full',
    [`spacing-vertical-${spacing[0]}`]: !!spacing[0],
    [`spacing-horizontal-${spacing[1]}`]: !!spacing[1],
    [`text-${textColor}`]: !!textColor,
    [`bg-${bgColor}`]: !!bgColor,
    [`border-${borderColor}`]: !!borderColor,
    [`border-${borderWidth}`]: !!borderWidth
  })

  const contentStyle = {
    ...style,
    ...(!!size[0] ? { width: size[0] } : {}),
    ...(!!size[1] ? { height: size[1] } : {})
  }

  // anchor
  if (href && !as) {
    return (
      <a href={href} className={containerClass} {...restProps}>
        <div className={contentClass} style={contentStyle}>
          {children}
        </div>
        <style jsx>{styles}</style>
      </a>
    )
  }

  // link
  if (href && as) {
    return (
      <Link href={href} as={as}>
        <a className={containerClass} {...restProps}>
          <div className={contentClass} style={contentStyle}>
            {children}
          </div>
          <style jsx>{styles}</style>
        </a>
      </Link>
    )
  }

  // button
  return (
    <button type={htmlType} className={containerClass} {...restProps}>
      <div className={contentClass} style={contentStyle}>
        {children}
      </div>
      <style jsx>{styles}</style>
    </button>
  )
}
