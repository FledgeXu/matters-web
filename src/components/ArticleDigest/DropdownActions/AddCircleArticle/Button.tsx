import { IconCircle24, Menu, TextIcon, Translate } from '~/components'

const AddCircleArticleButton = ({ openDialog }: { openDialog: () => void }) => {
  return (
    <Menu.Item onClick={openDialog}>
      <TextIcon icon={<IconCircle24 size="md" />} size="md" spacing="base">
        <Translate zh_hant="加入圍爐" zh_hans="加入围炉" en="Add to Circle" />
      </TextIcon>
    </Menu.Item>
  )
}

export default AddCircleArticleButton
