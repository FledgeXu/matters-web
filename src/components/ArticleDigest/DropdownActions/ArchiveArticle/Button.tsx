import { Icon, Menu, TextIcon, Translate } from '~/components'

const ArchiveArticleButton = ({ showDialog }: { showDialog: () => void }) => {
  return (
    <Menu.Item onClick={showDialog}>
      <TextIcon
        icon={<Icon.ArchiveMedium size="md" />}
        size="md"
        spacing="base"
      >
        <Translate zh_hant="站內隱藏" zh_hans="站内隐藏" />
      </TextIcon>
    </Menu.Item>
  )
}

export default ArchiveArticleButton
