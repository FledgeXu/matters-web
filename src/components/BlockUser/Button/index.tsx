import {
  IconMuteMedium,
  IconUnMuteMedium,
  Menu,
  TextIcon,
  Translate,
} from '~/components'
import { useMutation } from '~/components/GQL'
import TOGGLE_BLOCK_USER from '~/components/GQL/mutations/toggleBlockUser'

import { ADD_TOAST } from '~/common/enums'

import { ToggleBlockUser } from '~/components/GQL/mutations/__generated__/ToggleBlockUser'
import { BlockUserPrivate } from '../__generated__/BlockUserPrivate'
import { BlockUserPublic } from '../__generated__/BlockUserPublic'

const BlockUserButton = ({
  user,
  openDialog,
}: {
  user: BlockUserPublic & Partial<BlockUserPrivate>
  openDialog: () => void
}) => {
  const [unblockUser] = useMutation<ToggleBlockUser>(TOGGLE_BLOCK_USER, {
    variables: { id: user.id, enabled: false },
    optimisticResponse: {
      toggleBlockUser: {
        id: user.id,
        isBlocked: false,
        __typename: 'User',
      },
    },
  })
  const onUnblock = async () => {
    await unblockUser()
    window.dispatchEvent(
      new CustomEvent(ADD_TOAST, {
        detail: {
          color: 'green',
          content: <Translate id="successUnblock" />,
        },
      })
    )
  }

  if (user.isBlocked) {
    return (
      <Menu.Item onClick={onUnblock}>
        <TextIcon
          icon={<IconUnMuteMedium size="md" />}
          size="md"
          spacing="base"
        >
          <Translate id="unblockUser" />
        </TextIcon>
      </Menu.Item>
    )
  }

  return (
    <Menu.Item onClick={openDialog}>
      <TextIcon icon={<IconMuteMedium size="md" />} size="md" spacing="base">
        <Translate id="blockUser" />
      </TextIcon>
    </Menu.Item>
  )
}

export default BlockUserButton
