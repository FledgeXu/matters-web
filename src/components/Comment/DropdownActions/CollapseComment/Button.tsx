import { IconCollapse16, Menu, TextIcon, Translate } from '~/components'

const CollapseCommentButton = ({ openDialog }: { openDialog: () => void }) => {
  return (
    <Menu.Item onClick={openDialog}>
      <TextIcon icon={<IconCollapse16 size="md" />} size="md" spacing="base">
        <Translate id="collapseComment" />
      </TextIcon>
    </Menu.Item>
  )
}

export default CollapseCommentButton
