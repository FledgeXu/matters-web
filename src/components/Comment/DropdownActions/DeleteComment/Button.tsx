import { Icon, Menu, TextIcon, Translate } from '~/components'

const DeleteCommentButton = ({ showDialog }: { showDialog: () => void }) => {
  return (
    <Menu.Item onClick={showDialog}>
      <TextIcon icon={<Icon.RemoveMedium size="md" />} size="md" spacing="base">
        <Translate id="delete" />
      </TextIcon>
    </Menu.Item>
  )
}

export default DeleteCommentButton
