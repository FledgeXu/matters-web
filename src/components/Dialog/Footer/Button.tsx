import { forwardRef } from 'react'

import { Button, ButtonProps, IconSpinner, TextIcon } from '~/components'

type DialogFooterButtonProps = {
  textColor?: 'black' | 'white' | 'grey'
  bgColor?: 'green' | 'grey-lighter' | 'red' | 'white'
  icon?: React.ReactNode

  loading?: boolean
} & Omit<ButtonProps, 'bgColor'>

const DialogFooterButton: React.FC<DialogFooterButtonProps> = forwardRef(
  (
    {
      loading,
      textColor = 'white',
      bgColor = 'green',

      icon,

      disabled,
      children,
      ...restProps
    },
    ref
  ) => (
    <Button
      bgColor={bgColor}
      size={['100%', '3rem']}
      disabled={disabled || loading}
      ref={ref}
      {...restProps}
    >
      <TextIcon
        icon={icon || (loading && <IconSpinner size="md" />)}
        color={textColor}
        size="md"
        weight="md"
        textPlacement="left"
      >
        {loading ? null : children}
      </TextIcon>
    </Button>
  )
)

export default DialogFooterButton
