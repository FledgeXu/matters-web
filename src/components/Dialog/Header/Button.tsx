import { forwardRef } from 'react'

import { Button, ButtonProps, Icon, TextIcon, Translate } from '~/components'
import { useResponsive } from '~/components/Hook'

import { TEXT } from '~/common/enums'

interface CloseButtonProps {
  close: () => void
  ref: React.Ref<any>
}

export const CloseButton = forwardRef(({ close }: CloseButtonProps, ref) => {
  const isSmallUp = useResponsive({ type: 'sm-up' })()

  if (!process.browser) {
    return null
  }

  return (
    <Button
      onClick={close}
      aria-label="關閉"
      ref={ref}
      bgColor={isSmallUp ? 'grey-lighter' : undefined}
      bgHoverColor={isSmallUp ? 'green-lighter' : undefined}
      size={isSmallUp ? ['2rem', '2rem'] : undefined}
    >
      {!isSmallUp && (
        <TextIcon color="green" size="md">
          <Translate
            zh_hant={TEXT.zh_hant.cancel}
            zh_hans={TEXT.zh_hans.cancel}
          />
        </TextIcon>
      )}
      {isSmallUp && <Icon.CloseGreenMedium size="lg" />}
    </Button>
  )
})

type RightButtonProps = {
  text: string | React.ReactNode
  loading?: boolean
} & ButtonProps

export const RightButton: React.FC<RightButtonProps> = ({
  text,
  loading,
  ...buttonProps
}) => {
  const isSmallUp = useResponsive({ type: 'sm-up' })()

  if (!process.browser) {
    return null
  }

  return (
    <Button
      {...buttonProps}
      size={isSmallUp ? [null, '2rem'] : undefined}
      spacing={isSmallUp ? [0, 'base'] : undefined}
      bgColor={isSmallUp ? 'green' : undefined}
    >
      <TextIcon
        color={isSmallUp ? 'white' : 'green'}
        size={isSmallUp ? 'md-s' : 'md'}
        weight="md"
        icon={loading && <Icon.Spinner size="sm" />}
      >
        {loading ? null : text}
      </TextIcon>
    </Button>
  )
}
