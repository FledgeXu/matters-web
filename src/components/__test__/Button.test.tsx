import { render } from '@testing-library/react'
import React from 'react'

import { Button } from '../Button'

test('Button default style match snapshot', () => {
  const { container } = render(<Button>test</Button>)
  expect(container.firstChild).toMatchSnapshot()
})
