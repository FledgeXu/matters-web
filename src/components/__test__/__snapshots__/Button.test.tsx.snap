// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default style match snapshot 1`] = `
<button
  class="jsx-undefined btn size-default bg-white"
  type="button"
>
  <span
    class="jsx-undefined"
  >
    test
  </span>
</button>
`;
